src/responder.ts

```diff
@@ -150,83 +176,329 @@ export class EthereumDedicatedResponder extends EthereumResponder {
+        if (this.locked) {
+            throw new Error("This responder can ony handle one response at a time."); // TODO: more specific Error type?
+        }
+
+        const responseFlow = new EthereumResponseFlow(appointmentId, responseData);
+
+        this.setup();
+
+        const signerAddress = await promiseTimeout(
+            this.signer.getAddress(),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
+        // Get the current nonce to be used
+        this.nonce = await promiseTimeout(
+            this.signer.provider.getTransactionCount(signerAddress),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
+        // Get the initial gas price
+        this.gasPrice = await promiseTimeout(
+            this.signer.provider.getGasPrice(),
+            EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+        );
+
         while (this.mAttemptsDone < this.maxAttempts) {
             this.mAttemptsDone++;
             try {
                 // Try to call submitStateFunction, but timeout with an error if
-                // there is no response for 30 seconds.
-                const tx = await promiseTimeout(this.submitStateFunction(), 30000);
+                // there is no response for WAIT_TIME_FOR_PROVIDER_RESPONSE ms.
+                const tx = await promiseTimeout(
+                    this.submitStateFunction(responseData),
+                    EthereumDedicatedResponder.WAIT_TIME_FOR_PROVIDER_RESPONSE
+                );
 
                 // The response has been sent, but should not be considered confirmed yet.
-                this.responseFlow.status = ResponseState.ResponseSent;
-                this.asyncEmit(ResponderEvent.ResponseSent, this.responseFlow);
+                responseFlow.state = ResponseState.ResponseSent;
+                responseFlow.txHash = tx.hash;
+                this.asyncEmit(ResponderEvent.ResponseSent, responseFlow);
+
+
+                // Last block seen when transaction was first sent
+                const txSentBlockNumber = this.lastBlockNumberSeen;
+
+                // Promise that waits for the first confirmation
+                const firstConfirmationPromise = waitForConfirmations(this.signer.provider, tx.hash, 1);
+
+                // Promise that rejects after WAIT_BLOCKS_BEFORE_RETRYING blocks are mined since the transaction was first sent
+                const firstConfirmationTimeoutPromise = new Promise((_, reject) => {
+                    const testCondition = () => {
+                        if (this.lastBlockNumberSeen > txSentBlockNumber + EthereumDedicatedResponder.WAIT_BLOCKS_BEFORE_RETRYING) {
+                            reject(new StuckTransactionError(`Transaction still not mined after ${this.lastBlockNumberSeen - txSentBlockNumber} blocks`));
+                        }
+                        setTimeout(testCondition, 20);
```

Comment:
Just because they were written at different times. I'll put 1000 for both, as blocks don't come too often and the <1 sec response delay is not important.